text/0000-proc-macros.md

```diff
@@ -0,0 +1,417 @@
+passed to procedural macros does not need to satisfy the Rust parser, only the
+lexer. Stabilising an interface based on tokens means we need only commit to
+not changing the rules around those tokens, not the whole grammar. I.e., it
+allows us to change the Rust grammar without breaking procedural macros.
+
+In order to make the token-based interface even more flexible and future-proof,
+I propose a simpler token abstraction than is currently used in the compiler.
+The proposed system may be used directly in the compiler or may be an interface
+wrapper over a more efficient representation.
+
+Since macro expansion will not operate purely on tokens, we must keep hygiene
+information on tokens, rather than on `Ident` AST nodes (we might be able to
+optimise by not keeping such info for all tokens, but that is an implementation
+detail). We will also keep span information for each token, since that is where
+a record of macro expansion is maintained (and it will make life easier for
+tools. Again, we might optimise internally).
+
+A token is a single lexical element, for example, a numeric literal, a word
+(which could be an identifier or keyword), a string literal, or a comment.
+
+A token stream is a sequence of tokens, e.g., `a b c;` is a stream of four
+tokens - `['a', 'b', 'c', ';'']`.
+
+A token tree is a tree structure where each leaf node is a token and each
+interior node is a token stream. I.e., a token stream which can contain nested
+token streams. A token tree can be delimited, e.g., `a (b c);` will give
+`TT(None, ['a', TT(Some('()'), ['b', 'c'], ';'']))`. An undelimited token tree
+is useful for grouping tokens due to expansion, without representation in the
+source code. That could be used for unsafety hygiene, or to affect precedence
+and parsing without affecting scoping. They also replace the interpolated AST
+tokens currently in the compiler.
+
+In code:
+
+```
+// We might optimise this representation
+pub struct TokenStream(Vec<TokenTree>);
+
+// A borrowed TokenStream
+pub struct TokenSlice<'a>(&'a [TokenTree]);
+
+// A token or token tree.
+pub struct TokenTree {
+    pub kind: TokenKind,
+    pub span: Span,
+    pub hygiene: HygieneObject,
+}
+
+pub enum TokenKind {
+    Sequence(Delimiter, Vec<TokenTree>),
+
+    // The content of the comment can be found from the span.
+    Comment(CommentKind),
+    // The Span is the span of the string itself, without delimiters.
+    String(Span, StringKind),
+
+    // These tokens are treated specially since they are used for macro
+    // expansion or delimiting items.
+    Exclamation,  // `!`
+    Dollar,       // `$`
+    // Not actually sure if we need this or if semicolons can be treated like
+    // other punctuation.
+    Semicolon,    // `;`
+    Eof,
+
+    // Word is defined by Unicode Standard Annex 31 -
+    // [Unicode Identifier and Pattern Syntax](http://unicode.org/reports/tr31/)
+    Word(InternedString),
```

Comment:
Exposing `InternedString` makes me uneasy - the fact that we intern these is an implementation detail, although it does affect whether the API for accessing the contents involves `MacroContext`.

We could get away with either an interned index with the interner in TLS, or RC+SSO or some other strange combination, if we don't want to involve the `MacroContext` (which is the real decision to be made).

Still, this could probably use a less specific name, such as `Symbol` or even `Word`.